# Coding Exercises (Part 1)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## DataFrame Basics II

### Exercise 5: Filtering DataFrames & Adding/Removing Rows and Columns

In [ ]:
#run the cell!
import pandas as pd

In [ ]:
#run the cell!
cars = pd.read_csv("cars.csv")

In [ ]:
#run the cell!
cars

40. __Check__ for all elements in the column __"origin"__ whether they are __equal to "europe"__. __Save__ the boolean Series in the variable __mask1__! __Fill in the gaps__!

In [5]:
mask1 = cars.origin == 'europe'
mask1

0      False
1      False
2      False
3      False
4      False
5      False
6      False
7      False
8      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18     False
19      True
20      True
21      True
22      True
23      True
24     False
25     False
26     False
27     False
28     False
29     False
       ...  
368    False
369    False
370    False
371    False
372    False
373    False
374    False
375     True
376    False
377    False
378    False
379    False
380    False
381    False
382    False
383    False
384    False
385    False
386    False
387    False
388    False
389    False
390    False
391    False
392    False
393    False
394     True
395    False
396    False
397    False
Name: origin, Length: 398, dtype: bool

41. __Check__ for all elements in the column __"mpg"__ whether they are __smaller than 20__. __Save__ the boolean Series in the variable __mask2__!

In [6]:
mask2 = cars['mpg']<20

In [7]:
mask2

0       True
1       True
2       True
3       True
4       True
5       True
6       True
7       True
8       True
9       True
10      True
11      True
12      True
13      True
14     False
15     False
16      True
17     False
18     False
19     False
20     False
21     False
22     False
23     False
24     False
25      True
26      True
27      True
28      True
29     False
       ...  
368    False
369    False
370    False
371    False
372    False
373    False
374    False
375    False
376    False
377    False
378    False
379    False
380    False
381    False
382    False
383    False
384    False
385    False
386    False
387    False
388    False
389    False
390    False
391    False
392    False
393    False
394    False
395    False
396    False
397    False
Name: mpg, Length: 398, dtype: bool

42. __Filter__ the cars DataFrame for __all cars from europe__ (use mask1) and __save__ the subset in the variable __europe__! __Fill in the gaps!__

In [10]:
europe  = cars.name[mask1].copy()

__Inspect__! What is the name of the __second__ car?

In [12]:
# run the cell!
europe

19          vw 1131 deluxe sedan
20                   peugeot 504
21                   audi 100 ls
22                      saab 99e
23                      bmw 2002
50                     opel 1900
51                   peugeot 304
52                     fiat 124b
55                  vw model 111
59                     vw type 3
76               volvo 145e (sw)
77                   vw 411 (sw)
78              peugeot 504 (sw)
79               renault 12 (sw)
102              vw super beetle
114         fiat 124 sport coupe
117                     fiat 128
118                   opel manta
119                   audi 100ls
120                  volvo 144ea
122                    saab 99le
141                     audi fox
142                    vw dasher
143                   opel manta
147                     fiat 128
148                  fiat 124 tc
151                    fiat x1.9
172                    vw dasher
175                    vw rabbit
177                   audi 100ls
          

The second car is a ... peugeot 504!

43. Get some __meta information__ on the DataFrame europe. __How many__ cars are from europe? 

In [13]:
len(europe)

70

... 70 cars are from europe!

44. __Filter__ the DataFrame __cars__ for all cars __from europe with low fuel efficiency__ (mpg lower than 20). Use __mask1__ and __mask2__!<br> __Save__ the subset in the variable __europe_le__! __Fill in the gaps__!

In [ ]:
europe_le = cars.loc[---].copy()

__Inspect__! What is the __name__ of the __least efficient__ car from europe? 

In [ ]:
#run the cell!
europe_le

The least efficient european car is ... peugeot 604sl

45. __Filter__ the DataFrame cars for all cars with an __mpg between 10 and 15__ (both ends inclusive!). <br>
 __Save__ the subset in the variable __mpg_10_15__! __Fill in the gaps__!

In [16]:
mpg_10_15 = cars.loc[(cars['mpg']>=10) & (cars['mpg']<=15),:].copy()

__Inspect__! The __first__ car is...?

In [18]:
# run the cell!
mpg_10_15.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
1  15.0          8         350.0       165.0    3693          11.5   
5  15.0          8         429.0       198.0    4341          10.0   
6  14.0          8         454.0       220.0    4354           9.0   
7  14.0          8         440.0       215.0    4312           8.5   
8  14.0          8         455.0       225.0    4425          10.0   

   model_year origin               name  
1          70    usa  buick skylark 320  
5          70    usa   ford galaxie 500  
6          70    usa   chevrolet impala  
7          70    usa  plymouth fury iii  
8          70    usa   pontiac catalina

The __first__ car is... buick skylark 320!

__Inspect__! __How many__ cars are in the subset?

In [19]:
#run the cell!
mpg_10_15.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68 entries, 1 to 224
Data columns (total 9 columns):
mpg             68 non-null float64
cylinders       68 non-null int64
displacement    68 non-null float64
horsepower      68 non-null float64
weight          68 non-null int64
acceleration    68 non-null float64
model_year      68 non-null int64
origin          68 non-null object
name            68 non-null object
dtypes: float64(4), int64(3), object(2)
memory usage: 5.3+ KB


There are ... 68 cars!

46. __Filter__ the Dataframe __cars__ for all cars that are __not built in the years 73 and 74__, and only the columns __"mpg"__ and __"name"__!<br> __Save__ the subset in the variable __not_73_74__. __Fill in the gaps!__ 

In [ ]:
not_73_74 = cars.loc[~cars.model_year.---([73, 74]), ['mpg', 'name']].copy()

In [29]:
temp_no7374 = cars[~cars['model_year'].isin([73,74])]

In [30]:
temp_no7374

mpg  horsepower  weight  acceleration  model_year  origin  \
0    18.0       130.0    3504          12.0          70     usa   
1    15.0       165.0    3693          11.5          70     usa   
2    18.0       150.0    3436          11.0          70     usa   
3    16.0       150.0    3433          12.0          70     usa   
4    17.0       140.0    3449          10.5          70     usa   
5    15.0       198.0    4341          10.0          70     usa   
6    14.0       220.0    4354           9.0          70     usa   
7    14.0       215.0    4312           8.5          70     usa   
8    14.0       225.0    4425          10.0          70     usa   
9    15.0       190.0    3850           8.5          70     usa   
10   15.0       170.0    3563          10.0          70     usa   
11   14.0       160.0    3609           8.0          70     usa   
12   15.0       150.0    3761           9.5          70     usa   
13   14.0       225.0    3086          10.0          70     usa   
14   24.0        95.0    2372          15.0          70   japan   
15   22.0        95.0    2833          15.5          70     usa   
16   18.0        97.0    2774          15.5          70     usa   
17   21.0        85.0    2587          16.0          70     usa   
18   27.0        88.0    2130          14.5          70   japan   
19   26.0        46.0    1835          20.5          70  europe   
20   25.0        87.0    2672          17.5          70  europe   
21   24.0        90.0    2430          14.5          70  europe   
22   25.0        95.0    2375          17.5          70  europe   
23   26.0       113.0    2234          12.5          70  europe   
24   21.0        90.0    2648          15.0          70     usa   
25   10.0       215.0    4615          14.0          70     usa   
26   10.0       200.0    4376          15.0          70     usa   
27   11.0       210.0    4382          13.5          70     usa   
28    9.0       193.0    4732          18.5          70     usa   
29   27.0        88.0    2130          14.5          71   japan   
..    ...         ...     ...           ...         ...     ...   
368  27.0        88.0    2640          18.6          82     usa   
369  34.0        88.0    2395          18.0          82     usa   
370  31.0        85.0    2575          16.2          82     usa   
371  29.0        84.0    2525          16.0          82     usa   
372  27.0        90.0    2735          18.0          82     usa   
373  24.0        92.0    2865          16.4          82     usa   
374  23.0         NaN    3035          20.5          82     usa   
375  36.0        74.0    1980          15.3          82  europe   
376  37.0        68.0    2025          18.2          82   japan   
377  31.0        68.0    1970          17.6          82   japan   
378  38.0        63.0    2125          14.7          82     usa   
379  36.0        70.0    2125          17.3          82     usa   
380  36.0        88.0    2160          14.5          82   japan   
381  36.0        75.0    2205          14.5          82   japan   
382  34.0        70.0    2245          16.9          82   japan   
383  38.0        67.0    1965          15.0          82   japan   
384  32.0        67.0    1965          15.7          82   japan   
385  38.0        67.0    1995          16.2          82   japan   
386  25.0       110.0    2945          16.4          82     usa   
387  38.0        85.0    3015          17.0          82     usa   
388  26.0        92.0    2585          14.5          82     usa   
389  22.0       112.0    2835          14.7          82     usa   
390  32.0        96.0    2665          13.9          82   japan   
391  36.0        84.0    2370          13.0          82     usa   
392  27.0        90.0    2950          17.3          82     usa   
393  27.0        86.0    2790          15.6          82     usa   
394  44.0        52.0    2130          24.6          82  europe   
395  32.0        84.0    2295          11.6          

In [ ]:
#run the cell!
not_73_74

__Inspect__! __How many__ cars are in this subset?

In [ ]:
# run the cell!
not_73_74.info()

There are ... 331 cars in the subset! 

In [ ]:
# run the cell!
cars

47. __Drop__ the columns __"displacement"__ and __"cylinders"__ from the DataFrame cars! __Save__ the change!

In [20]:
cars.drop(['displacement','cylinders'],axis=1,inplace=True)

In [21]:
# run the cell!
cars

mpg  horsepower  weight  acceleration  model_year  origin  \
0    18.0       130.0    3504          12.0          70     usa   
1    15.0       165.0    3693          11.5          70     usa   
2    18.0       150.0    3436          11.0          70     usa   
3    16.0       150.0    3433          12.0          70     usa   
4    17.0       140.0    3449          10.5          70     usa   
5    15.0       198.0    4341          10.0          70     usa   
6    14.0       220.0    4354           9.0          70     usa   
7    14.0       215.0    4312           8.5          70     usa   
8    14.0       225.0    4425          10.0          70     usa   
9    15.0       190.0    3850           8.5          70     usa   
10   15.0       170.0    3563          10.0          70     usa   
11   14.0       160.0    3609           8.0          70     usa   
12   15.0       150.0    3761           9.5          70     usa   
13   14.0       225.0    3086          10.0          70     usa   
14   24.0        95.0    2372          15.0          70   japan   
15   22.0        95.0    2833          15.5          70     usa   
16   18.0        97.0    2774          15.5          70     usa   
17   21.0        85.0    2587          16.0          70     usa   
18   27.0        88.0    2130          14.5          70   japan   
19   26.0        46.0    1835          20.5          70  europe   
20   25.0        87.0    2672          17.5          70  europe   
21   24.0        90.0    2430          14.5          70  europe   
22   25.0        95.0    2375          17.5          70  europe   
23   26.0       113.0    2234          12.5          70  europe   
24   21.0        90.0    2648          15.0          70     usa   
25   10.0       215.0    4615          14.0          70     usa   
26   10.0       200.0    4376          15.0          70     usa   
27   11.0       210.0    4382          13.5          70     usa   
28    9.0       193.0    4732          18.5          70     usa   
29   27.0        88.0    2130          14.5          71   japan   
..    ...         ...     ...           ...         ...     ...   
368  27.0        88.0    2640          18.6          82     usa   
369  34.0        88.0    2395          18.0          82     usa   
370  31.0        85.0    2575          16.2          82     usa   
371  29.0        84.0    2525          16.0          82     usa   
372  27.0        90.0    2735          18.0          82     usa   
373  24.0        92.0    2865          16.4          82     usa   
374  23.0         NaN    3035          20.5          82     usa   
375  36.0        74.0    1980          15.3          82  europe   
376  37.0        68.0    2025          18.2          82   japan   
377  31.0        68.0    1970          17.6          82   japan   
378  38.0        63.0    2125          14.7          82     usa   
379  36.0        70.0    2125          17.3          82     usa   
380  36.0        88.0    2160          14.5          82   japan   
381  36.0        75.0    2205          14.5          82   japan   
382  34.0        70.0    2245          16.9          82   japan   
383  38.0        67.0    1965          15.0          82   japan   
384  32.0        67.0    1965          15.7          82   japan   
385  38.0        67.0    1995          16.2          82   japan   
386  25.0       110.0    2945          16.4          82     usa   
387  38.0        85.0    3015          17.0          82     usa   
388  26.0        92.0    2585          14.5          82     usa   
389  22.0       112.0    2835          14.7          82     usa   
390  32.0        96.0    2665          13.9          82   japan   
391  36.0        84.0    2370          13.0          82     usa   
392  27.0        90.0    2950          17.3          82     usa   
393  27.0        86.0    2790          15.6          82     usa   
394  44.0        52.0    2130          24.6          82  europe   
395  32.0        84.0    2295          11.6          

48. __Drop__ all rows/cars from __"usa"__ from the DataFrame cars! Use the __loc__ operator and __overwrite cars__!

In [ ]:
# run the cell!
cars

49. __Add__ the new column __"l_per_100km"__ to the DataFrame cars by calculating __235.21/mpg__. __Round__ to 2 decimals. __Fill in the gaps__!

In [22]:
cars['I_per_100'] = (235.21/cars['mpg']).round(2)

__Inspect__ cars! What is the __"l_per_100km"__ value for the __audi 100 ls__?

In [28]:
# run the cell!
cars[cars['name']=='audi 100 is']

Empty DataFrame
Columns: [mpg, horsepower, weight, acceleration, model_year, origin, name, I_per_100]
Index: []

The "l_per_100km" value for the audi 100 ls is ... 9.80!

# Well Done!

---------------------------------

# Hints (Spoiler!)

40. origin attribute; "europe"

41. mask2  =  cars.---  <  --- 

42. loc operator, mask1

43. info() method

44. Condition1 & Condition2

45. between() method

46. isin() method and ~ (tilde symbol)

47. drop() method

48. use != (not equal)

49. cars["new_column_label"]